packages/turf-line-intersect/index.ts

```diff
@@ -35,100 +35,56 @@ function lineIntersect<
   G2 extends LineString | MultiLineString | Polygon | MultiPolygon
 >(
   line1: FeatureCollection<G1> | Feature<G1> | G1,
-  line2: FeatureCollection<G2> | Feature<G2> | G2
+  line2: FeatureCollection<G2> | Feature<G2> | G2,
+  options: {
+    removeDuplicates?: boolean;
+  } = {}
 ): FeatureCollection<Point> {
-  const unique: any = {};
-  const results: any[] = [];
-
-  // First, normalize geometries to features
-  // Then, handle simple 2-vertex segments
-  if (line1.type === "LineString") {
-    line1 = feature(line1);
+  let removeDuplicates = true;
+  if ("removeDuplicates" in options) {
+    removeDuplicates = options.removeDuplicates;
   }
-  if (line2.type === "LineString") {
-    line2 = feature(line2);
+
+  let features = [];
+  if (line1.type === "FeatureCollection")
+    features = features.concat(line1.features);
+  else if (line1.type === "Feature") features.push(line1);
+  else if (
+    line1.type === "LineString" ||
+    line1.type === "Polygon" ||
+    line1.type === "MultiLineString" ||
+    line1.type === "MultiPolygon"
+  ) {
+    features.push(feature(line1));
   }
-  if (
-    line1.type === "Feature" &&
-    line2.type === "Feature" &&
-    line1.geometry !== null &&
-    line2.geometry !== null &&
-    line1.geometry.type === "LineString" &&
-    line2.geometry.type === "LineString" &&
-    line1.geometry.coordinates.length === 2 &&
-    line2.geometry.coordinates.length === 2
+
+  if (line2.type === "FeatureCollection")
+    features = features.concat(line2.features);
+  else if (line2.type === "Feature") features.push(line2);
+  else if (
+    line2.type === "LineString" ||
+    line2.type === "Polygon" ||
+    line2.type === "MultiLineString" ||
+    line2.type === "MultiPolygon"
   ) {
-    const intersect = intersects(line1, line2);
-    if (intersect) {
-      results.push(intersect);
-    }
-    return featureCollection(results);
+    features.push(feature(line2));
   }
 
-  // Handles complex GeoJSON Geometries
-  const tree = rbush();
-  tree.load(lineSegment(line2));
-  featureEach(lineSegment(line1), (segment) => {
-    featureEach(tree.search(segment), (match) => {
-      const intersect = intersects(segment, match);
-      if (intersect) {
-        // prevent duplicate points https://github.com/Turfjs/turf/issues/688
-        const key = getCoords(intersect).join(",");
-        if (!unique[key]) {
-          unique[key] = true;
-          results.push(intersect);
-        }
+  const intersections = findIntersections(featureCollection(features));
```

Comment:
This seems like it works differently than the previous version. Before it would compare line1 to line2, but now if you have a FeatureCollection for line1 it will be intersected with its own geometries?

Reply:
Hmm yes good pickup on a subtle difference I hadn't thought of - this new implementation will find self-intersections in a geometry, as well as intersections between two geometries. I could possibly add something to the `sweepline-intersections` lib to handle that if we thought that was problematic.

Reply:
Its a breaking change in a way that isn't really possible to even work around as a library consumer. That seems pretty bad.

Reply:
I've updated my sweepline-intersections library to allow libraries to specify whether they want to detect self-intersections or not, and made the subsequent change in how we use it here in Turf. So I think this is good to go now.